modules/heat_conduction/src/actions/MortarGapHeatTransferAction.C

```diff
@@ -0,0 +1,349 @@
+
+InputParameters
+MortarGapHeatTransferAction::validParams()
+{
+  InputParameters params = Action::validParams();
+  params.addClassDescription(
+      "Action that controls the creation of all of the necessary objects for "
+      "calculation of heat transfer through an open/closed gap using a mortar formulation and a "
+      "modular design approach");
+
+  params.addParam<Real>("thermal_lm_scaling",
+                        1.,
+                        "Scaling factor to apply to the thermal Lagrange multiplier variable");
+
+  params += ModularGapConductanceConstraint::validParams();
+  params += GapFluxModelRadiation::validParams();
+  params += GapFluxModelConduction::validParams();
+
+  params.makeParamNotRequired<SubdomainName>("primary_subdomain");
+  params.makeParamNotRequired<SubdomainName>("secondary_subdomain");
+  params.makeParamNotRequired<Real>("gap_conductivity");
+  params.addParam<bool>(
+      "user_provided_user_objects",
+      true,
+      "Whether user creates user objects in the input file and provides their block names");
+
+  params.addParam<MooseEnum>("gap_flux_options",
+                             MortarGapHeatTransferAction::getModelEnum(),
+                             "The gap flux models to build");
+
+  params.addParam<std::vector<UserObjectName>>(
+      "user_created_user_object_physics",
+      "The name of the user objects created by the user to represent gap heat transfer physics");
+
+  return params;
+}
+
+MortarGapHeatTransferAction::MortarGapHeatTransferAction(const InputParameters & params)
+  : Action(params),
+    _user_provided_mortar_meshes(false),
+    _user_provided_user_objects(getParam<bool>("user_provided_user_objects")),
+    _gap_flux_models(getParam<MooseEnum>("gap_flux_options").getEnum<UserObjectToBuild>())
+
+{
+
+  if (params.isParamSetByUser("user_provided_user_objects"))
+    _existing_user_objects =
+        getParam<std::vector<UserObjectName>>("user_created_user_object_physics");
+  else if (_user_provided_user_objects)
+    paramError("user_created_user_object_physics",
+               "User selected to provided user created user objects in the input file, but fail to "
+               "provide the user object names");
+
+  if (params.isParamSetByUser("primary_subdomain") &&
+      params.isParamSetByUser("secondary_subdomain"))
+  {
+    mooseInfo("Mortar gap heat transfer action is using the lower-dimensional domains provided by "
+              "the user");
+    _user_provided_mortar_meshes = true;
+  }
+  else
+    mooseInfo("Mortar gap heat transfer action is creating new lower-dimensional domains");
+}
+
+void
+MortarGapHeatTransferAction::act()
+{
+  if (_current_task == "append_mesh_generator")
+    addMortarMesh();
+  else if (_current_task == "add_mortar_variable")
+    addMortarVariable();
+  if (_current_task == "add_constraint")
+    addConstraints();
+  else if (_current_task == "add_user_object")
+    if (!_user_provided_user_objects)
```

Comment:
Why not let the user take advantage of both options. E.g. one might want to add 

```
gap_flux_options = 'RADIATION'
user_provided_user_objects = 'my_pressure_dependent_conduction'
```

Reply:
True. However, if a user wants to add his/her own physics via tailored user objects, I can't see much benefit by letting construct other general user objects (describing more basic physics) in the action. This behavior seems obscure to me, but logically easy to implement.